### PR TITLE
Handle skeleton encoding internally

### PR DESCRIPTION
**Pull request by eberrigan**
### Description

Looking at our [`fly_skeleton_legs.json` test data](https://github.com/talmolab/sleap/blob/3c7f5afa8dd952fdee8ef3c8ac94b0689dac9cdb/tests/data/skeleton/fly_skeleton_legs.json):
Our final `json_str` has keys `directed`, `graph`, `links`, `multigraph`, `nodes`.
   - `directed` is a boolean. _(UNCHANGED from input graph)_
   - `graph` is a dict with keys `name` and `num_edges_inserted`. _(UNCHANGED from input graph)_
   - `links` is a list of dicts with keys `edge_insert_idx`, `key`, `source`, `target`, `type`, for each edge.
   - `multigraph` is a boolean = True. _(UNCHANGED from input graph)_
   - `nodes` is a list of dicts with keys `id` for each node.

| input object to JSON encoder | python object location in graph dict        | encoded object as JSON string                                                                             |
|------------------------------------|----------------------------------------------------|-----------------------------------------------------------------------------------------------------------------|
|`Node(name='neck', weight=1.0)`     |     `"links":[{... source: { Node(name='neck', weight=1.0) }`     |     `{"py/object": "sleap.skeleton.Node", "py/state": {"py/tuple": ["neck", 1.0]}}`     |
| `Node(name='head', weight=1.0) `     |    `"links":[{... target: { Node(name='head', weight=1.0) }`      |      `{"py/object": "sleap.skeleton.Node", "py/state": {"py/tuple": ["head", 1.0]}}`     |
| `<EdgeType.BODY: 1>`               | `"links":[{... type: { <EdgeType.BODY: 1> }`      | `{"py/reduce": [{"py/type": "sleap.skeleton.EdgeType"}, {"py/tuple": [1]}, null, null, null]}`        |
| `Node(name='head', weight=1.0)`     |     `"nodes": [{'id': Node(name='head', weight=1.0)}, ..., ]`     |     `{"id": {"py/id": 2}}`     |

- This is because `Skeleton._graph` is passed through `json_graph.node_link_data`, which returns a dictionary with node-link formatted data (see documentation [here](https://networkx.org/documentation/stable/reference/readwrite/generated/networkx.readwrite.json_graph.node_link_data.html). 
```
def node_link_data():
     multigraph = G.is_multigraph()

    # Allow 'key' to be omitted from attrs if the graph is not a multigraph.
    key = None if not multigraph else key
    if len({source, target, key}) < 3:
        raise nx.NetworkXError("Attribute names are not unique.")
    data = {
        "directed": G.is_directed(),
        "multigraph": multigraph,
        "graph": G.graph,
        "nodes": [{**G.nodes[n], name: n} for n in G],
    }
    if multigraph:
        data[link] = [
            {**d, source: u, target: v, key: k}
            for u, v, k, d in G.edges(keys=True, data=True)
        ]
    else:
        data[link] = [{**d, source: u, target: v} for u, v, d in G.edges(data=True)]
    return data
```
 - **Note** that our node-linked data has key `links` and not `link`.
      -  We do not have `networkx` [pinned](https://github.com/talmolab/sleap/blob/3c7f5afa8dd952fdee8ef3c8ac94b0689dac9cdb/.conda/meta.yaml#L39). I am not sure why that hasn't been a problem.
- Then the node-link formatted data is passed to `jsonpickle.encode(data)`. 
     - Serializes the SLEAP class `Node`. Each [`Node`](https://github.com/talmolab/sleap/blob/3c7f5afa8dd952fdee8ef3c8ac94b0689dac9cdb/sleap/skeleton.py#L49) object, which has attributes `name` and `weight`, is encoded as a Python object `sleap.skeleton.Node`. The `Node`'s internal state is represented as a Python tuple containing the `name` (as a string) and `weight` (as a numerical value): `["name", weight]`.
          - In the case where the node is an int, such as when the skeleton has a `node_to_idx` mapping (https://github.com/talmolab/sleap/blob/3c7f5afa8dd952fdee8ef3c8ac94b0689dac9cdb/sleap/skeleton.py#L1002-L1006), i.e. when the skeleton is from `Labels`,  the node is returned as an integer without the name and weight attributes.
     - Serializes the SLEAP class `EdgeType`. [`EdgeType`](https://github.com/talmolab/sleap/blob/3c7f5afa8dd952fdee8ef3c8ac94b0689dac9cdb/sleap/skeleton.py#L31) inherits from Python's `Enum` class. The possible edge types are `BODY = 1` and `SYMMETRY = 2`. When serialized, `EdgeType` is represented using Python's `reduce` function, storing the type as `sleap.skeleton.EdgeType` and the value as a Python tuple containing the `Enum` value.
- If the object has been "seen" before, it will not be encoded as the full JSON string but referenced by its `py/id`, which starts at 1 and indexes the objects in the order they are seen so that the second time the first object is used, it will be referenced as `{"py/id": 1}`.
     - This is tricky because looking at [fly_skeleton_legs.json](https://github.com/talmolab/sleap/blob/3c7f5afa8dd952fdee8ef3c8ac94b0689dac9cdb/tests/data/skeleton/fly_skeleton_legs.json) we can see that the first time a node is used in a `source` or `target` in the `links` list, it is encoded as a py/object with a py/state. Then it is referenced later on using the py/id. `nodes` then uses only the py/id to reference the `Node`. 
     - However this is not what you would expect from looking at the order in of the graph passed from `jsongraph.node_link_data()` to `jsonpickle.encode()` (shown below) which has `nodes` first and `links` after. 
     - The same is true for the order of `source`, `target` and `type`: in the graph below `type` comes before `source` and `target` so unless explicitly ordered after `source` and `target` the py/id will not be consistent with the legacy data. 
     - For this reason, a method is added to the `SkeletonEncoder` called `_encode_links` that is used within the classmethod `encode` to encode the `links` first, and within the `links` list, encode the `source`, `target` and then the `type`.
     - If the exact order wasn't followed, and the py/ids were not the same as the legacy data, the `EdgeType` would be decoded as a `Node` and the edges would not be formed correctly in the decoded `Skeleton` using [`Skeleton.from_json`](https://github.com/talmolab/sleap/blob/3c7f5afa8dd952fdee8ef3c8ac94b0689dac9cdb/sleap/skeleton.py#L1057-L1087). 
```
{'directed': True, 'multigraph': True, 'graph': {'name': 'skeleton_legs.mat', 'num_edges_inserted': 23}, 'nodes': [{'id': Node(name='head', weight=1.0)}, {'id': Node(name='neck', weight=1.0)}, {'id': Node(name='thorax', weight=1.0)}, {'id': Node(name='abdomen', weight=1.0)}, {'id': Node(name='wingL', weight=1.0)}, {'id': Node(name='wingR', weight=1.0)}, {'id': Node(name='forelegL1', weight=1.0)}, {'id': Node(name='forelegL2', weight=1.0)}, {'id': Node(name='forelegL3', weight=1.0)}, {'id': Node(name='forelegR1', weight=1.0)}, {'id': Node(name='forelegR2', weight=1.0)}, {'id': Node(name='forelegR3', weight=1.0)}, {'id': Node(name='midlegL1', weight=1.0)}, {'id': Node(name='midlegL2', weight=1.0)}, {'id': Node(name='midlegL3', weight=1.0)}, {'id': Node(name='midlegR1', weight=1.0)}, {'id': Node(name='midlegR2', weight=1.0)}, {'id': Node(name='midlegR3', weight=1.0)}, {'id': Node(name='hindlegL1', weight=1.0)}, {'id': Node(name='hindlegL2', weight=1.0)}, {'id': Node(name='hindlegL3', weight=1.0)}, {'id': Node(name='hindlegR1', weight=1.0)}, {'id': Node(name='hindlegR2', weight=1.0)}, {'id': Node(name='hindlegR3', weight=1.0)}], 'links': [{'edge_insert_idx': 1, 'type': <EdgeType.BODY: 1>, 'source': Node(name='neck', weight=1.0), 'target': Node(name='head', weight=1.0), 'key': 0}, {'edge_insert_idx': 0, 'type': <EdgeType.BODY: 1>, 'source': Node(name='thorax', weight=1.0), 'target': Node(name='neck', weight=1.0), 'key': 0}, {'edge_insert_idx': 2, 'type': <EdgeType.BODY: 1>, 'source': Node(name='thorax', weight=1.0), 'target': Node(name='abdomen', weight=1.0), 'key': 0}, {'edge_insert_idx': 3, 'type': <EdgeType.BODY: 1>, 'source': Node(name='thorax', weight=1.0), 'target': Node(name='wingL', weight=1.0), 'key': 0}, {'edge_insert_idx': 4, 'type': <EdgeType.BODY: 1>, 'source': Node(name='thorax', weight=1.0), 'target': Node(name='wingR', weight=1.0), 'key': 0}, {'edge_insert_idx': 5, 'type': <EdgeType.BODY: 1>, 'source': Node(name='thorax', weight=1.0), 'target': Node(name='forelegL1', weight=1.0), 'key': 0}, {'edge_insert_idx': 8, 'type': <EdgeType.BODY: 1>, 'source': Node(name='thorax', weight=1.0), 'target': Node(name='forelegR1', weight=1.0), 'key': 0}, {'edge_insert_idx': 11, 'type': <EdgeType.BODY: 1>, 'source': Node(name='thorax', weight=1.0), 'target': Node(name='midlegL1', weight=1.0), 'key': 0}, {'edge_insert_idx': 14, 'type': <EdgeType.BODY: 1>, 'source': Node(name='thorax', weight=1.0), 'target': Node(name='midlegR1', weight=1.0), 'key': 0}, {'edge_insert_idx': 17, 'type': <EdgeType.BODY: 1>, 'source': Node(name='thorax', weight=1.0), 'target': Node(name='hindlegL1', weight=1.0), 'key': 0}, {'edge_insert_idx': 20, 'type': <EdgeType.BODY: 1>, 'source': Node(name='thorax', weight=1.0), 'target': Node(name='hindlegR1', weight=1.0), 'key': 0}, {'edge_insert_idx': 6, 'type': <EdgeType.BODY: 1>, 'source': Node(name='forelegL1', weight=1.0), 'target': Node(name='forelegL2', weight=1.0), 'key': 0}, {'edge_insert_idx': 7, 'type': <EdgeType.BODY: 1>, 'source': Node(name='forelegL2', weight=1.0), 'target': Node(name='forelegL3', weight=1.0), 'key': 0}, {'edge_insert_idx': 9, 'type': <EdgeType.BODY: 1>, 'source': Node(name='forelegR1', weight=1.0), 'target': Node(name='forelegR2', weight=1.0), 'key': 0}, {'edge_insert_idx': 10, 'type': <EdgeType.BODY: 1>, 'source': Node(name='forelegR2', weight=1.0), 'target': Node(name='forelegR3', weight=1.0), 'key': 0}, {'edge_insert_idx': 12, 'type': <EdgeType.BODY: 1>, 'source': Node(name='midlegL1', weight=1.0), 'target': Node(name='midlegL2', weight=1.0), 'key': 0}, {'edge_insert_idx': 13, 'type': <EdgeType.BODY: 1>, 'source': Node(name='midlegL2', weight=1.0), 'target': Node(name='midlegL3', weight=1.0), 'key': 0}, {'edge_insert_idx': 15, 'type': <EdgeType.BODY: 1>, 'source': Node(name='midlegR1', weight=1.0), 'target': Node(name='midlegR2', weight=1.0), 'key': 0}, {'edge_insert_idx': 16, 'type': <EdgeType.BODY: 1>, 'source': Node(name='midlegR2', weight=1.0), 'target': Node(name='midlegR3', weight=1.0), 'key': 0}, {'edge_insert_idx': 18, 'type': <EdgeType.BODY: 1>, 'source': Node(name='hindlegL1', weight=1.0), 'target': Node(name='hindlegL2', weight=1.0), 'key': 0}, {'edge_insert_idx': 19, 'type': <EdgeType.BODY: 1>, 'source': Node(name='hindlegL2', weight=1.0), 'target': Node(name='hindlegL3', weight=1.0), 'key': 0}, {'edge_insert_idx': 21, 'type': <EdgeType.BODY: 1>, 'source': Node(name='hindlegR1', weight=1.0), 'target': Node(name='hindlegR2', weight=1.0), 'key': 0}, {'edge_insert_idx': 22, 'type': <EdgeType.BODY: 1>, 'source': Node(name='hindlegR2', weight=1.0), 'target': Node(name='hindlegR3', weight=1.0), 'key': 0}]}
```

- A test has been added which loads a skeleton from a JSON file, get the graph with `json_graph.node_link_data` from `networkx`, encodes the graph with the new `SkeletonEncoder.encode()` method, then uses [`Skeleton.from_json`](https://github.com/talmolab/sleap/blob/3c7f5afa8dd952fdee8ef3c8ac94b0689dac9cdb/sleap/skeleton.py#L1057-L1087) to deserialize the JSON string (with `jsonpickle.decode`, then `json_graph.node_link_graph`). The [`Skeleton.matches()`](https://github.com/talmolab/sleap/blob/3c7f5afa8dd952fdee8ef3c8ac94b0689dac9cdb/sleap/skeleton.py#L153-L179) method is used to test the equivalence of the deserialized `Skeletons`. 
     [X] This test could be parametrized to do with various skeleton fixtures.
          [X] We should make sure to do this with a template fixture at least.
     [X] An additional test could be added to see if the serialized JSON strings are equivalent.
     - **Important**: This test currently fails when `attrs` is unconstrained ([see github run](https://github.com/talmolab/sleap/actions/runs/11007387249)) because with the version of `attrs` pulled in, the nodes are  `json_pickle` fails to decode the `Node`s which have been encoded as tuples by our `SkeletonEncoder.encode()` (to maintain backwards compatibility).
          - """The change in attrs version used by jsonpickle to write/read the skeleton.json is causing problems, namely, attrs >=22.2.0 returns the object state as a dictionary instead of a tuple. This results in incompatibilities when attempting to deserialize skeleton.json files (w/ an older attrs <22.2.0) that were serialized with attrs >=22.2.0. Based on the discussion in https://github.com/python-attrs/attrs/pull/1085 it sounds like attrs ==22.2.0 was unable to deserialize pickled data from attrs <22.2.0, and attrs ==23.1.0 (>=23.1.0?) at least fixes backwards compatibility. The long-term plan for attrs (discussed in https://github.com/python-attrs/attrs/issues/1091) is to remove support for deserializing tuples after giving everyone a few versions of attrs to convert their data over to a dictionary serialization.""" from https://github.com/talmolab/sleap/issues/1918 .

### Types of changes

- [ ] Bugfix
- [ ] New feature
- [X] Refactor / Code style update (no logical changes)
- [ ] Build / CI changes
- [ ] Documentation Update
- [ ] Other (explain)

### Does this address any currently open issues?
#1470 #1918 

This pull request accompanies #1961 in handling JSON encoding and decoding internally instead of relying on `jsonpickle`.

### Outside contributors checklist

- [ ] Review the [guidelines for contributing](https://github.com/talmolab/sleap/blob/develop/docs/CONTRIBUTING.md) to this repository
- [ ] Read and sign the [CLA](https://github.com/talmolab/sleap/blob/develop/sleap-cla.pdf) and add yourself to the [authors list](https://github.com/talmolab/sleap/blob/develop/AUTHORS)
- [ ] Make sure you are making a pull request against the **develop** branch (not *main*). Also you should start *your branch* off *develop*
- [ ] Add tests that prove your fix is effective or that your feature works
- [ ] Add necessary documentation (if appropriate)

#### Thank you for contributing to SLEAP!
:heart:

## Summary by CodeRabbit

- **New Features**
	- Introduced a new `SkeletonEncoder` class for enhanced encoding of skeleton representations.
	- Added methods for encoding `Node` and `EdgeType` objects into JSON formats.

- **Chores**
	- Updated version constraints for the `attrs` package in environment configuration files.

- **Tests**
	- Added new tests for validating the encoding and decoding of `Skeleton` objects.

### CODE REVIEW

File: sleap/skeleton.py [hunk 1]
```diff
@@ -1020,10 +1230,15 @@ def to_json(self, node_to_idx: Optional[Dict[Node, int]] = None) -> str:
...
                 "preview_image": self.preview_image,
             }
+            print(f"data: {data}")
         else:
             data = graph
+            print(f"data: {data}")
 
-        json_str = jsonpickle.encode(data)
+        # json_str = jsonpickle.encode(data)
+        encoder = SkeletonEncoder()
```
Comment (eberrigan) [inline sleap/skeleton.py:1239]: `SkeletonEncoder.encode(data)` instead

File: sleap/skeleton.py [hunk 2]
```diff
@@ -376,6 +376,194 @@ def decode(cls, json_str: str) -> Dict:
...
+
+        Returns:
+            A list of encoded edge dictionaries with keys ordered as specified.
+        """
+        encoded_links = []
+        for link in links:
+            # Use a regular dict (insertion order preserved in Python 3.7+)
+            encoded_link = {}
+
+            # Encode in specific order of appearance
```
Comment (roomrys) [inline sleap/skeleton.py:470]:
```suggestion
            
```

Comment (roomrys) [inline sleap/skeleton.py:470]:
```suggestion


```

**Review by roomrys — APPROVED** [review]

**Comment by roomrys** [comment]
Related:
- #1841

**Pull request merged**